data_safe_haven/infrastructure/programs/sre/firewall.py

```diff
@@ -111,206 +114,238 @@ def __init__(
+                        name="AllowUbuntuSnapcraft",
+                        protocols=[
+                            network.AzureFirewallApplicationRuleProtocolArgs(
+                                port=int(Ports.HTTPS),
+                                protocol_type=network.AzureFirewallApplicationRuleProtocolType.HTTPS,
+                            ),
+                        ],
+                        source_addresses=props.subnet_workspaces_prefixes,
+                        target_fqdns=PermittedDomains.UBUNTU_SNAPCRAFT,
+                    ),
+                    network.AzureFirewallApplicationRuleArgs(
+                        description="Allow external RStudio deb downloads",
+                        name="AllowRStudioDeb",
+                        protocols=[
+                            network.AzureFirewallApplicationRuleProtocolArgs(
+                                port=int(Ports.HTTPS),
+                                protocol_type=network.AzureFirewallApplicationRuleProtocolType.HTTPS,
+                            ),
+                        ],
+                        source_addresses=props.subnet_workspaces_prefixes,
+                        target_fqdns=PermittedDomains.RSTUDIO_DEB,
+                    ),
+                ],
+            ),
+            network.AzureFirewallApplicationRuleCollectionArgs(
+                action=network.AzureFirewallRCActionArgs(
+                    type=network.AzureFirewallRCActionType.DENY
                 ),
-                network.AzureFirewallApplicationRuleCollectionArgs(
+                name="workspaces-deny",
+                priority=FirewallPriorities.SRE_WORKSPACES_DENY,
+                rules=[
+                    network.AzureFirewallApplicationRuleArgs(
+                        description="Deny external Ubuntu Snap Store upload and login access",
+                        name="DenyUbuntuSnapcraft",
+                        protocols=[
+                            network.AzureFirewallApplicationRuleProtocolArgs(
+                                port=int(Ports.HTTP),
+                                protocol_type=network.AzureFirewallApplicationRuleProtocolType.HTTP,
+                            ),
+                            network.AzureFirewallApplicationRuleProtocolArgs(
+                                port=int(Ports.HTTPS),
+                                protocol_type=network.AzureFirewallApplicationRuleProtocolType.HTTPS,
+                            ),
+                        ],
+                        source_addresses=props.subnet_workspaces_prefixes,
+                        target_fqdns=ForbiddenDomains.UBUNTU_SNAPCRAFT,
+                    ),
+                ],
+            ),
+        ]
+
+        network_rule_collections: list[
+            network.AzureFirewallNetworkRuleCollectionArgs
+        ] = []
+
+        if props.allow_workspace_internet:
+            application_rule_collections = []
```

Comment:
It feels strange to be to do it this way,

```python
list_a = [...]
list_b = [...]

if condition:
    list_a = []
    list_b.append(...)
```

I think it might be clearer like

```python
list_b_common = [...]

if condition:
   list_a = []
   list_b = list_b_common.append(...)
else:
   list_a = [...]
   list_b = list_b_common
```